<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [502]:
import pandas as pd
import psycopg2

In [503]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [504]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [505]:
# текст запроса
query_3_1 = f'''select count(*) 
            from vacancies
            limit {10}
        '''

In [506]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\2706544688.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [507]:
# текст запроса
query_3_2 = f'''select count(*) 
            from employers
        '''

In [508]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\3359072823.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [509]:
# текст запроса
query_3_3 = f'''select count(*) 
            from areas
        '''

In [510]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\3170201742.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [511]:
# текст запроса
query_3_4 = f'''select count(*) 
            from industries
        '''

In [512]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\978673360.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

Из предварительного анализа данных мы можем сделать выводы, что:  
1. Вакансий у нас больше, чем работадателей. Это значит, что часто бывает такое, что один работодатель размещает несколько вакансий.   
2. Данный сервис охватывает большое количество регионов и сфер деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [513]:
# текст запроса
query_4_1 = f'''select a.name area, count(*) cnt
            from areas a 
            join vacancies v on a.id = v.area_id
            group by 1
            order by 2 desc
            limit 5
        '''

In [514]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\2947587922.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [515]:
# текст запроса
query_4_2 = f'''select count(*)
            from vacancies
            where salary_from is not null or salary_to is not null
        '''

In [516]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\1502081924.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [517]:
# текст запроса
query_4_3 = f'''select round(avg(salary_from),0) lower, round(avg(salary_to),0) upper
            from vacancies
        '''

In [518]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\318603332.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


lower     upper
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [519]:
# текст запроса
query_4_4 = f'''select schedule, employment, count(*)
            from vacancies
            group by 1, 2
            order by 3 desc
        '''

In [520]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\1605305061.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [521]:
# текст запроса
query_4_5 = f'''select experience, count(*)
            from vacancies
            group by 1
            order by 2
        '''

In [522]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\117783236.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

Мы можем сделать следующие выводы из детального анализа вакансий:  
1. Почти половина работодателей не указывет зарплату.
2. Самой востребованной комбинацей графика и занятости является 'Полный день - Полная занятость'. Вероятно, что работодатели хотят, чтобы сотрудники больше работали.
3. Самый востребованный опыт работы - от 1 до 3 лет

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [523]:
# текст запроса
query_5_1 = f'''select e.name, count(*)
            from vacancies v
            join employers e on v.employer_id = e.id
            group by 1
            order by 2 desc
            limit 5
        '''

In [524]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\2071789291.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [525]:
# текст запроса
query_5_2 = f'''select a.name area,
            count(distinct e.id) employers,
            count(distinct v.id) vacancies
            from areas a
                left join employers e on a.id = e.area
                left join vacancies v on v.area_id = a.id
            where v.id is null
            group by a.id
            order by 2 desc
        '''

In [526]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\1661138368.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


area  employers  vacancies
0                Россия        410          0
1             Казахстан        207          0
2    Московская область         75          0
3    Краснодарский край         19          0
4    Ростовская область         18          0
..                  ...        ...        ...
588             Скидель          0          0
589          Белоозерск          0          0
590           Орловский          0          0
591          Павловская          0          0
592             Таксимо          0          0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [527]:
# текст запроса
query_5_3 = f'''select e.name, count(distinct v.area_id)
            from employers e, vacancies v, areas a
            where v.area_id = a.id and v.employer_id = e.id
            group by e.id
            order by 2 desc
        '''

In [528]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\3759115417.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  count
0                                      Яндекс    181
1                                  Ростелеком    152
2                                  Спецремонт    116
3                      Поляков Денис Иванович     88
4                                   ООО ЕФИН      71
...                                       ...    ...
14901                            НПП Авиатрон      1
14902              Центр дистанционных торгов      1
14903  Городские Телекоммуникационные Системы      1
14904                       Введенский, Отель      1
14905              Группа Компаний Автоимпорт      1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [529]:
# текст запроса
query_5_4 = f'''select count(e.id)
            from employers e
            left join employers_industries i on e.id = i.employer_id
            where i.industry_id is null
            group by employer_id
        '''

In [530]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\1388507732.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [531]:
# текст запроса
query_5_5 = f'''select e.name
            from employers e
            left join employers_industries i on e.id = i.employer_id
            group by e.id
            having count(i.industry_id) = 4
            order by 1
            offset 2
            limit 1
        '''

In [532]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\2612825567.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [533]:
# текст запроса
query_5_6 = f'''select count(*)
            from employers_industries e
            left join industries i on i.id = e.industry_id
            where i.name = 'Разработка программного обеспечения'
        '''

In [534]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\3088500550.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [535]:
# код для получения списка городов-милионников
import requests # Импортируем библиотеку requests
from bs4 import BeautifulSoup # Импортируем библиотеку BeautifulSoup
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России' # Определяем адрес страницы
response = requests.get(url) # Выполняем GET-запрос, содержимое ответа присваивается переменной response
page = BeautifulSoup(response.text, 'html.parser') # Создаём объект BeautifulSoup, указывая html-парсер
links = page.select('td a') # Ищем все ссылки на странице и сохраняем в переменной links в виде списка
cities =[]
for link in links: # добавляем города в список
    cities.append(link.text)
cities = cities[1:17] # убираем не города из списка
print(cities)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']


In [536]:
# текст запроса
query_5_7 = f'''select a.name city, count(v.id) vac_cnt
            from vacancies v
                left join employers e on v.employer_id = e.id
                left join areas a on v.area_id = a.id
            where a.name in {tuple(cities)} and e.name = 'Яндекс'
            group by a.name
            union all
            select 'total', count(*)
            from vacancies v
                left join employers e on v.employer_id = e.id
                left join areas a on v.area_id = a.id
            where a.name in {tuple(cities)} and e.name = 'Яндекс'
            order by 2
        '''

In [537]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\613699941.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


city  vac_cnt
0              Омск       21
1         Челябинск       22
2        Красноярск       23
3         Волгоград       24
4             Пермь       25
5            Казань       25
6    Ростов-на-Дону       25
7               Уфа       26
8            Самара       26
9         Краснодар       30
10          Воронеж       32
11      Новосибирск       35
12  Нижний Новгород       36
13     Екатеринбург       39
14  Санкт-Петербург       42
15           Москва       54
16            total      485

***

Выводы по анализу работодателей:  
1. Больше всего вакансий на hh.ru размещено у компании Яндекс.
2. У одной компании часто бывает несколько сфер деятельности.
3. Больше всего вакансий находятся в Москве.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [538]:
# текст запроса
query_6_1 = f'''select count(*)
            from vacancies
            where lower(name) like '%data%' or lower(name) like '%данн%'
        '''

In [539]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\1850917348.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [540]:
# текст запроса
query_6_2 = f'''select count(*)
            from vacancies
            where (lower(name) like '%data scientist%' or lower(name) like '%data science%' 
            or lower(name) like '%исследователь данных%'
            or (lower(name) like '%ml%' and lower(name) not like '%html%')
            or lower(name) like '%machine learning%'
            or lower(name) like '%машинн%обучен%') and
            (lower(name) like '%junior%' 
            or experience = 'Нет опыта'
            or employment = 'Стажировка')
        '''

In [541]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\3060336679.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [542]:
# текст запроса
query_6_3 = f'''select count(*)
            from vacancies
            where (lower(name) like '%data scientist%' or lower(name) like '%data science%' 
            or lower(name) like '%исследователь данных%'
            or (name like '%ML%' and lower(name) not like '%html%')
            or lower(name) like '%machine learning%'
            or lower(name) like '%машинн%обучен%') and
            (lower(key_skills) like '%sql%' or lower(key_skills) like '%postgres%')
        '''

In [543]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\3046237206.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [544]:
# текст запроса
query_6_4 = f'''select count(*)
            from vacancies
            where (lower(name) like '%data scientist%' or lower(name) like '%data science%' 
            or lower(name) like '%исследователь данных%'
            or (name like '%ML%' and lower(name) not like '%html%')
            or lower(name) like '%machine learning%'
            or lower(name) like '%машинн%обучен%') and
            lower(key_skills) like '%python%'
        '''

In [545]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\1859741363.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [546]:
# текст запроса
query_6_5 = f'''select round(avg(length(key_skills) - length(replace(key_skills, CHR(9), ''))+1), 2)
            from vacancies
            where lower(name) like '%data%scientist%' or lower(name) like '%data%science%' 
            or lower(name) like '%исследователь%данных%'
            or (name like '%ML%' and lower(name) not like '%html%')
            or lower(name) like '%machine%learning%'
            or lower(name) like '%машинн%обучен%'
        '''

In [547]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\4252439027.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [548]:
# текст запроса
query_6_5 = f'''select experience, round(avg(coalesce((vacancies.salary_from + vacancies.salary_to) / 2, vacancies.salary_from, vacancies.salary_to)), 0)
            from vacancies
            where (lower(name) like '%data%scientist%' or lower(name) like '%data%science%' 
            or lower(name) like '%исследователь%данных%'
            or (name like '%ML%' and lower(name) not like '%html%')
            or lower(name) like '%machine%learning%'
            or lower(name) like '%машинн%обучен%')
            and (salary_from is not null 
            or salary_to is not null)
            group by experience
        '''

In [549]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\4252439027.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


experience     round
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

***

Выводы по предметному анализу:  
1. Вакансий, связанных с данными 1771, что достаточно немного, учитывая, что всего их 49197.
2. В вакансиях для дата-сайентиста часто указываются SQL и Python в ключевых навыках, но при этом Python указывается чаще.
3. Чем больше опыт, тем больше средняя зарплата.
4. Для вакансий, где необходим опыт работы более 6 лет часто ен указывается зарплата.

# Дополнительные исследования

Найдем сколько всего вакансий есть для дата-сайентиста (вне зависимости от уровня) и в каких сферах они востребованы.

In [550]:
# текст запроса
query_new = f'''select count(*)
            from vacancies
            where lower(name) like '%data%scientist%' or lower(name) like '%data%science%' 
            or lower(name) like '%исследователь%данных%'
            or (name like '%ML%' and lower(name) not like '%html%')
            or lower(name) like '%machine%learning%'
            or lower(name) like '%машинн%обучен%'
        '''

In [551]:
# результат запроса
df = pd.read_sql_query(query_new, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\3169162569.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_new, connection)


count
0    483

In [552]:
# текст запроса
query_new_2 = f'''select i.name, count(*)
            from employers_industries e
            left join vacancies v on e.employer_id = v.employer_id
            left join industries i on e.industry_id = i.id
            where lower(v.name) like '%data%scientist%' or lower(v.name) like '%data%science%' 
            or lower(v.name) like '%исследователь%данных%'
            or (v.name like '%ML%' and lower(v.name) not like '%html%')
            or lower(v.name) like '%machine%learning%'
            or lower(v.name) like '%машинн%обучен%'
            group by i.id
            order by 2 desc
        '''

In [553]:
# результат запроса
df = pd.read_sql_query(query_new_2, connection)
df 

C:\Users\evgus\AppData\Local\Temp\ipykernel_28160\387136983.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_new_2, connection)


name  count
0                 Разработка программного обеспечения    148
1   Системная интеграция,  автоматизации технологи...    139
2                                                Банк     93
3   Интернет-компания (поисковики, платежные систе...     79
4                                     Мобильная связь     23
..                                                ...    ...
92  Бытовая химия, парфюмерия, косметика  (продвиж...      1
93  Контроль качества, экспертиза, испытания и сер...      1
94  Бытовая химия, парфюмерия, косметика  (произво...      1
95  Добыча и обогащение минерального сырья (соль, ...      1
96  Теплоэнергетика (генерация электроэнергии и те...      1

[97 rows x 2 columns]

Мы можем увидеть, что самые популярные сферы деятельности, где нужны дата-сайентисты это сферы, связанные с итернетом и новыми технологиями, такими, как:
1. Разработка программного обеспечения
2. Банк
3. Мобильная связь 
4. И т.д.  

Но реже встречаются и такие сферы деятельности, которые на первый взгляд никак не связаны с data science, например:
1. Бытовая химия, парфюмерия, косметика
2. Контроль качества, экспертиза
3. Добыча и обогащение минерального сырья
4. И т.д.  

Это может означать, что дата-сайентист может пригодится в любой сфере деятельности.

# Общий вывод по проекту

Лидерами по размещенным вакансиям являются крупные компании, предоставляющие технологические финансовые и интернет-услуги.
Чаще всего появляются вакансии связанные с программированием и администрацией. Средняя зарплатная вилка находится в промежутке от 71000 до 11000.  

Больше всего вакансий находится в Москве и Санкт-Питербурге. Но вопрос с регионами хорошо было бы изучить дополнительно, т.к. в таблице с регионами указаны не только города, но и страны. Например как регион указаны не только города России, но и сама Россия. Также в данном проекте мы не полностью изучили распределение количества вакансий по всем городам, взяв только города-миллионники. 

Также я бы провела дополнительные исследования вакансий для дата-сайентистов и вывела бы самые популярные профессиональные навыки, которыми должен обладать дата-сайентист. Кроме того, если бы у нас в таблице был бы столбец с количеством откликов на вакансию, интересно бы было узнать, как зависит количество откликлв от указания зарплаты или от указания сферы деятельности.